### ISSUE TITLE
Occasional "broken pipe" errors on Travis/AppVeyor during reflection test

### ISSUE THREAD

**Issue by tkelman** [JuliaLang/julia#9501]
These tests were added in #8251 - I've been seeing intermittent failures on both Travis and AppVeyor over the last month or so (https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.946/job/ih37028wuuky5mvs and https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.67/job/dyhnx0wobs5l5kkj and a bunch more I'll have to go digging for).

AppVeyor almost always gives an error message like
```
	From worker 3:	     * reflection
ERROR: test error in expression: code_native(ismatch,(Regex,AbstractString)) == nothing
write: broken pipe (EPIPE)
 in wait at no file
 in stream_wait at no file
while loading reflection.jl, in expression starting on line 11
while loading C:\projects\julia\test\runtests.jl, in expression starting on line 42
```
but often on Travis there won't be an error, it'll just manifest as the worker that's running the reflection test dying.

The I/O label is due to this code at the start of `test/reflection.jl`:
```
# redirect stdout and stderr to avoid spam during tests.
oldout = STDOUT
olderr = STDERR
redirect_stdout()
redirect_stderr()
```

**Comment by timholy** [comment]
It sure seems like a race condition. I can strip `runtests.jl` down to this:
```julia
testnames = [
    "linalg", "core", "keywordargs", "numbers", "strings", "dates",
    "collections", "hashing", "remote", "iobuffer", "staged", "arrayops",
    "subarray", "reduce", "reducedim", "random", "intfuncs",
    "simdloop", "blas", "fft", "dsp", "sparse", "bitarray", "copy", "math",
    "functional", "bigint", "sorting", "statistics", "spawn",
    "backtrace", "priorityqueue", "arpack", "file", "suitesparse", "version",
    "resolve", "pollfd", "mpfr", "broadcast", "complex", "socket",
    "floatapprox", "readdlm", "reflection", "regex", "float16", "combinatorics",
    "sysinfo", "rounding", "ranges", "mod2pi", "euler", "show",
    "lineedit", "replcompletions", "repl", "test", "goto",
    "llvmcall", "grisu", "nullable", "meta", "profile",
    "libgit2", "docs", "base64"
]

tests = ARGS

@everywhere include("testdefs.jl")
# include("testdefs.jl")

reduce(propagate_errors, nothing, pmap(runtests, tests; err_retry=false, err_stop=true))

println("    \033[32;1mSUCCESS\033[0m")
```
and still get the 141 exit code much of the time. Curiously, if I delete the definition of `testnames` (which is unused), I almost never get the 141 code. And if I comment out the `@everywhere` line and uncomment the one below it, I almost never get the 141 code.

I don't know enough about pipes, etc to take this any farther.

**Comment by timholy** [comment]
Hmm, somehow this got automatically closed when #9678 was merged. Reopening.

**Comment by tkelman** [comment]
Whoops, it had the string "fix #9501" in the PR description.

**Comment by SimonDanisch** [comment]
Is this related?
I get this randomly in some of my test code:
```Julia
error in running finalizer: jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
jl_uv_writecb() ERROR: broken pipe EPIPE
```

**Comment by SimonDanisch** [comment]
It seems to stem from an error in the finalizer, which then wants to be printed but i guess the pipe is already freed? So not really related, but also not very nice...

**Comment by vtjnash** [comment]
resolved by #9521, the original bug was in the test (redirecting stdout, but not reading from it to prevent kernel from blocking on more data)
